system/service/auth_credentials.go

```diff
@@ -0,0 +1,1157 @@
+package service
+
+// part of auth service
+// collection of functions that handle password checking, setting, resetting, changing
+//
+// general credential handling functions should still be part of auth.go
+
+import (
+	"context"
+	"fmt"
+	internalAuth "github.com/cortezaproject/corteza-server/pkg/auth"
+	"github.com/cortezaproject/corteza-server/pkg/errors"
+	"github.com/cortezaproject/corteza-server/pkg/filter"
+	"github.com/cortezaproject/corteza-server/pkg/rand"
+	"github.com/cortezaproject/corteza-server/store"
+	"github.com/cortezaproject/corteza-server/system/types"
+	"github.com/dgryski/dgoogauth"
+	"golang.org/x/crypto/bcrypt"
+	"math"
+	rand2 "math/rand"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+)
+
+const (
+	credentialsTypePassword                    = "password"
+	credentialsTypePersistentSession           = "persistent-session"
+	credentialsTypeEmailAuthToken              = "email-authentication-token"
+	credentialsTypeResetPasswordToken          = "password-reset-token"
+	credentialsTypeResetPasswordTokenExchanged = "password-reset-token-exchanged"
+	credentialsTypeCreatePasswordToken         = "password-create-token"
+	credentialsTypeMfaTotpSecret               = "mfa-totp-secret"
+	credentialsTypeMFAEmailOTP                 = "mfa-email-otp"
+
+	credentialsTokenLength = 32
+
+	tokenReqMaxCount  = 5
+	tokenReqMaxWindow = time.Minute * 15
+
+	passwordMinLength = 8
+	passwordMaxLength = 256
+)
+
+var (
+	oneTokenPerUser = map[string]bool{
+		credentialsTypeResetPasswordToken: true,
+	}
+)
+
+// ValidateEmailConfirmationToken issues a validation token that can be used for
+func (svc *auth) ValidateEmailConfirmationToken(ctx context.Context, token string) (user *types.User, err error) {
+	return svc.loadFromTokenAndConfirmEmail(ctx, token, credentialsTypeEmailAuthToken)
+}
+
+// loadFromTokenAndConfirmEmail loads token, confirms user's
+func (svc *auth) loadFromTokenAndConfirmEmail(ctx context.Context, token, tokenType string) (u *types.User, err error) {
+	var (
+		aam = &authActionProps{
+			user:        u,
+			credentials: &types.Credentials{Kind: tokenType},
+		}
+	)
+
+	err = func() error {
+		if !svc.settings.Auth.Internal.Enabled {
+			return AuthErrInternalSignupDisabledByConfig(aam)
+		}
+
+		u, err = svc.loadUserFromToken(ctx, token, tokenType)
+		if err != nil {
+			return err
+		}
+
+		aam.setUser(u)
+		ctx = internalAuth.SetIdentityToContext(ctx, u)
+
+		if !u.EmailConfirmed {
+			// User's email is not confirmed but going through password reset flow
+			// we can confirm it
+			u.EmailConfirmed = true
+			u.UpdatedAt = now()
+			if err = store.UpdateUser(ctx, svc.store, u); err != nil {
+				return err
+			}
+		}
```

Comment:
Worth considering having another event here; `on userEmailConfirmed` for things like welcome or success email notifications, lead -> customer conversion, ...